include/deal.II/lac/parpack_solver.h

```diff
@@ -815,192 +830,159 @@ void PArpackSolver<VectorType>::solve
-              dst.extract_subvector_to (local_indices.begin(),
+              system_matrix.vmult(tmp, src);
+              // store M*X in X
+              tmp.extract_subvector_to (local_indices.begin(),
                                         local_indices.end(),
-                                        &workd[0]+shift_y  );
-
+                                        &workd[0]+shift_x);
+              inverse.vmult(dst,tmp);
             }
-            break;
-
-            case  2:
-              // compute  Y = B * X  where
-              // IPNTR(1) is the pointer into WORKD for X,
-              // IPNTR(2) is the pointer into WORKD for Y.
+          else if (mode == 1)
             {
+              system_matrix.vmult(dst, src);
+            }
+          else
+            AssertThrow (false, PArpackExcMode(mode));
 
-              const int shift_x = ipntr[0]-1;
-              const int shift_y = ipntr[1]-1;
-              Assert (shift_x>=0, dealii::ExcInternalError() );
-              Assert (shift_x+nloc <= (int)workd.size(), dealii::ExcInternalError() );
-              Assert (shift_y>=0, dealii::ExcInternalError() );
-              Assert (shift_y+nloc <= (int)workd.size(), dealii::ExcInternalError() );
-
-              src = 0.0;
-              src.add (nloc,
-                       &local_indices[0],
-                       &workd[0]+shift_x );
-              src.compress (VectorOperation::add);
-
-              // Multiplication with mass matrix M
-              mass_matrix.vmult(dst, src);
-
-              // store the result
-              dst.extract_subvector_to (local_indices.begin(),
-                                        local_indices.end(),
-                                        &workd[0]+shift_y);
+        }
+      else if (ido == 1 && mode >= 3)
+        // compute  Y = OP * X for mode 3, 4 and 5, where
+        // the vector B * X is already available in WORKD(ipntr(3)).
+        {
+          const int shift_b_x = ipntr[2]-1;
+          Assert (shift_b_x>=0, dealii::ExcInternalError() );
+          Assert (shift_b_x+nloc <= (int)workd.size(), dealii::ExcInternalError() );
+
+          // B*X
+          src.add (nloc,
+                   &local_indices[0],
+                   &workd[0]+shift_b_x );
+          src.compress (VectorOperation::add);
+
+          // solving linear system
+          Assert (mode == 3, ExcNotImplemented());
+          inverse.vmult(dst,src);
+        }
+      else if (ido == 2)
+        // compute  Y = B * X
+        {
+          src.add (nloc,
+                   &local_indices[0],
+                   &workd[0]+shift_x );
+          src.compress (VectorOperation::add);
 
+          // Multiplication with mass matrix M
+          if (mode == 1)
+            {
+              dst = src;
             }
-            break;
-
-            default:
-              AssertThrow (false, PArpackExcIdo(ido));
-              break;
+          else if (mode != 4)
+            // mode 2,3 and 5 have B=M
```

Comment:
this looks odd since you don't currently support modes 4, 5. So the condition is always true.

This place may call for using a `switch` statement on `mode`. If you want to proactively already support a case that's not supported elsewhere (e.g., because it's so simple -- like it is here -- that it's not worth excluding these modes), then maybe add a comment about it. Otherwise, it looks like a bug.